### DISCUSSION TITLE
arch: arm: Using ISR_DIRECT_PM with zero-latency-interrupt violation

### DISCUSSION THREAD

**Discussion by ibaz-nordic** [zephyrproject-rtos/zephyr#50720] [General]
**Describe the bug**
Calling ISR_DIRECT_PM when using ZLI will lead to [kernel API](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm/core/aarch32/irq_manage.c#L183).  This violates [ZLI requirement](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/kernel/services/interrupts.html#zero-latency-interrupts)

I've tried removing ISR_DIRECT_PM but this gives unstable behavior.  Is there another solution?

**Comment by mmahadevan108** [comment]
cc @carlescufi

**Reply by carlescufi** [reply]
I think this one is for @gmarull, @andyross  and @cvinayak

**Reply by carlescufi** [reply]
ZLIs in general cannot interact with the kernel at all, so I doubt that you can call `ISR_DIRECT_PM` at all from a ZLI.

**Reply by cvinayak** [reply]
There is also this: https://github.com/zephyrproject-rtos/zephyr/issues/32342 which I had asked as part of the PR review rework. Basically, using CONFIG_PM would use `cpsid i` to disable all interrupts:
https://github.com/zephyrproject-rtos/zephyr/blob/4b91c2d79fcc08abc8f8dce7cdcab9ea9fba2d0b/arch/arm/core/aarch32/isr_wrapper.S#L151-L169

Basically, I would not use `CONFIG_PM` if using `CONFIG_ZERO_LATENCY_IRQS=y`, then
https://github.com/zephyrproject-rtos/zephyr/blob/5b3f80094ec561758bd506c20c453703b859b8c6/include/zephyr/arch/arm/aarch32/irq.h#L143-L148

Zephyr Bluetooth Low Energy Controller does not use ZLI by default due to the reason of ZLI not being truely ZLI with `CONFIG_PM=y`.
